### ISSUE TITLE
RNNForward  -> Feature Not Implemented

### ISSUE THREAD

**Issue by simpleoier** [Microsoft/CNTK#1380]
When I test a BiLSTM model with CNTK using CPU (trained using GPU), an error occurs:
```
Inside File: Source/Math/Matrix.cpp  Line: 4423  Function: RNNForward  -> Feature Not Implemented.
Closed Kaldi writer.
```

Then I turned to test the model using GPU, no more such error.

Has anyone met the same error?

**Comment by dongyu888** [comment]
This is because the RNN node used is based on cuDNN and is only implemented in GPU.

 

 

Thanks,

 

Dong

**Comment by bencherian** [comment]
I don't think that RNN's are only supported on the GPU...there are examples that specifically mention using the CPU...that wouldn't make sense if they were entirely unsupported.

**Comment by tangyuq** [comment]
I am closing this issue for now. Please feel free to re-open it if you have further comments.

**Comment by ChangLiangLiuu** [comment]
I got this error when I do model evaluation with C++ API function void CNTK::Function::Evaluate(). I do this on CPU. It is still not implemented on CPU?

**Comment by pegulhane** [comment]
I am still seeing this  issue on my cntk 2.4 on windows. when will it be fixed.

**Comment by j-a-lewis** [comment]
I encountered the issue.  If you look within the CNTK ops module, the method causing problem (for me) resides at following package location and has the following comment (in part)  : ...\site-packages\cntk\ops\__init__.py

def optimized_rnnstack(operand, weights, hidden_size, num_layers,
                       bidirectional=False, recurrent_op='lstm', name=''):
    '''
    An RNN implementation that uses the primitives in cuDNN.
    If cuDNN is not available it fails. You can use :class:`~cntk.misc.optimized_rnnstack_converter.convert_optimized_rnnstack`
    to convert a model to GEMM-based implementation when no cuDNN.
.... '''

Where required I used the original method call to C.optimized_rnnstack as the argument for a call to suggested GEMM-based RNN model via the method C.misc.optimized_rnnstack_converter.convert_optimized_rnnstack, i.e.

> return C.misc.optimized_rnnstack_converter.convert_optimized_rnnstack(C.optimized_rnnstack(operand, weights=W, hidden_size=hidden_size, num_layers=num_layers, bidirectional=True, recurrent_op='lstm' ))

Although BLSTM fitting proceeded in very pokey fashion on my CPU-only Win10 machine (2.5 hrs/epoch), fitting over 10 epochs gave as approximately as good a fit as using a DNN feedforward network over 100 epochs (and I didn't have to buy a machine with an nVidia GPU!).
![20200205 BLSTM FIt on CPU-Only Win10 Machine](https://user-images.githubusercontent.com/41112084/73817495-a8cc5100-47b0-11ea-8396-b15344315f97.png)
